README

About
-----

Read and manipulate data in the GPX format. (see http://www.topografix.com/GPX/ for more).

Uses Processing's XML parser so only basic information (lat/lon/ele/time) is parsed.

A GPX file is an xml file with a collection of GPX Tracks inside.  Each GPX Track has a name, and is a collection of GPX Track Segments, and each segment is a collection of GPX Points.  Each point has a latitude and longitude as well as an elevation and timestamp.  No other information is parsed by this library.  A GPX file can also contain waypoints, with lat/lon position and name and type.

This library has only been tested with a few different files, and I expect it will need updating to properly handle things like waypoints with elevation, and so on.  If you find a file that breaks it, please send it to me.

Usage
-----

import tomc.gpx.*;

// outside setup()
GPX gpx;

  // inside setup()
  gpx = new GPX(this);

  // when you want to load data
  gpx.parse("test.gpx"); // or a URL

  // inside draw()
  for (int i = 0; i < gpx.getTrackCount(); i++) {
    GPXTrack trk = gpx.getTrack(i);
    // do something with trk.name
    for (int j = 0; j < trk.size(); j++) {
      GPXTrackSeg trkseg = trk.getTrackSeg(j);
      for (int k = 0; k < trkseg.size(); k++) {
        GPXPoint pt = trkseg.getPoint(k);
        // do something with pt.lat or pt.lon
      }
    }
  }

  for (int i = 0; i < gpx.getWayPointCount(); i++) {
    GPXWayPoint wpt = gpx.getWayPoint(i);
    // do something with wpt.lat or wpt.lon or wpt.name or wpt.type
  }


Installation
------------

Place this folder (gpx) into your Processing/libraries folder, then restart Processing.
Then choose GPX from the library import options in the Processing application menu.
If it's working, "import tomc.gpx" should be added to the top of your application file.


Bugs / Requests
---------------

Please file an issue at https://github.com/RandomEtc/processing-gpx/issues


Enjoy! -- TomC.



CHANGES
09/02/2013 13:27 PST
--------------------

Updated for Processing 2.0+, closes #2.

01/06/2011 10:33 PDT
--------------------

Added to github, fixed up a few organisational things.
Removed nanoxml and moved to Processing's own XML parser.


10/03/2007 10:40 CST
--------------------

Updated for Processing 0123+ (PApplet.reader became PApplet.createReader)


10/02/2006 09:55 GMT
--------------------

Initial release on processinghacks.com.




GPXPoint.java
/*

  Copyright (c) 2006-2011 Tom Carden

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

*/

package tomc.gpx;

import java.util.Date;
import java.text.SimpleDateFormat;
import processing.data.XML;

/** simple GPX point, only understands latitude, longitude, elevation and time
    FIXME extend parsing to understand other properties */
public class GPXPoint {

  private static final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'hh:mm:ss'Z'");
  private static final SimpleDateFormat msdateFormat = new SimpleDateFormat("yyyy-MM-dd'T'hh:mm:ss.SSS'Z'"); // allow millis too

  public double lat,lon;
  public double ele;
  public Date time;

  public GPXPoint() {
    this.lat = 0.0;
    this.lon = 0.0;
    this.ele = 0.0;
    this.time = new Date();
  }

  public GPXPoint(double lat, double lon, double ele, Date time) {
    this.lat = lat;
    this.lon = lon;
    this.ele = ele;
    this.time = time;
  }

  /** parses a GPXPoint from the given XML
      understands 2 time formats: ISO 8601 with and without milliseconds */
  public GPXPoint(XML trkpt) {

    // NB:- this is a bit more complex than it should be
    //      but it should handle a wider variety of broken
    //      GPX files, and let's face it that's why you're using
    //      a library instead of parsing this file yourself

    try {
      String sLat = trkpt.getString("lat");
      this.lat = Double.parseDouble(sLat);
    }
    catch (Exception e) {
      if (GPX.debug) {
        e.printStackTrace();
        System.err.println("error parsing lat in: " + trkpt);
      }
      this.lat = 0.0;
    }
    try {
      String sLon = trkpt.getString("lon");
      this.lon = Double.parseDouble(sLon);
    }
    catch (Exception e) {
      if (GPX.debug) {
        e.printStackTrace();
        System.err.println("error parsing lon in: " + trkpt);
      }
      this.lon = 0.0;
    }

    for (XML element : trkpt.getChildren()) {
      if (element.getName().equals("ele")) {
        try {
          String sEle = element.getContent();
          this.ele = Double.parseDouble(sEle);
        }
        catch (Exception e) {
          if (GPX.debug) {
            e.printStackTrace();
            System.err.println("error parsing ele in: " + trkpt);
          }
          this.ele = 0.0;
        }
      }
      else if (element.getName().equals("time")) {
        String t = element.getContent();
        try {
          if (t.length() == 20) {
            this.time = dateFormat.parse(t);
          }
          else if (t.length() == 24) {
            this.time = msdateFormat.parse(t);
          }
          else {
             this.time = new Date(Long.parseLong(t)); // try for unix time
    	  }
        }
        catch(Exception e) {
          if (GPX.debug) {
            e.printStackTrace();
            System.err.println("error parsing time in: " + trkpt);
          }
          this.time = new Date();
        }
      }
      else {
        if (GPX.debug) {
          System.err.print("skipping unknown element: ");
          System.err.println(element.getName());
        }
      }
    }

  }

}


GPXTrack.java
/*

  Copyright (c) 2006-2011 Tom Carden

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

*/

package tomc.gpx;

import processing.data.XML;
import java.util.Vector;

/** a collection of GPXTrackSeg objects, with type-safe convenience methods for adding/removing/getting track segments */
public class GPXTrack extends Vector {

  public String name;

  public GPXTrack(XML xmltrk) {
    this.name = xmltrk.getChild("name").getContent();
    for (XML child : xmltrk.getChildren("trkseg")) {
      this.addTrackSeg(new GPXTrackSeg(child));
    }
  }

  public GPXTrackSeg getTrackSeg(int i) {
    return (GPXTrackSeg)elementAt(i);
  }

  public void addTrackSeg(GPXTrackSeg trkseg) {
    addElement(trkseg);
  }

  public void removeTrackSeg(GPXTrackSeg trkseg) {
    removeElement(trkseg);
  }

}


GPXTrackSeg.java
/*

  Copyright (c) 2006-2011 Tom Carden

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

*/

package tomc.gpx;

import processing.data.XML;
import java.util.Vector;

/** a collection of GPXPoint objects, with type-safe convenience methods for adding/removing/getting points */
public class GPXTrackSeg extends Vector {

  public GPXTrackSeg(XML trkseg) {
    for (XML child : trkseg.getChildren("trkpt")) {
      addPoint(new GPXPoint(child));
    }
  }

  public void addPoint(GPXPoint trkpt) {
    addElement(trkpt);
  }

  public GPXPoint getPoint(int i) {
    return (GPXPoint)elementAt(i);
  }

  public void removePoint(GPXPoint trkpt) {
    removeElement(trkpt);
  }

}


GPXWayPoint.java
/*

  Copyright (c) 2006-2011 Tom Carden

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

*/

package tomc.gpx;

import processing.data.XML;

/** simple GPX waypoint, only understands latitude, longitude, name and type
    FIXME extend parsing to understand other properties */
public class GPXWayPoint {

  public double lat,lon;
  public String name;
  public String type;

  public GPXWayPoint() {
    this.lat = 0.0;
    this.lon = 0.0;
    this.name = "";
    this.type = "";
  }

  public GPXWayPoint(double lat, double lon, String name, String type) {
    this.lat = lat;
    this.lon = lon;
    this.name = name;
    this.type = type;
  }

  /** parses a GPXWayPoint from the given XML */
  public GPXWayPoint(XML trkpt) {
    this.lat = trkpt.getDouble("lat");
    this.lon = trkpt.getDouble("lon");
    this.name = trkpt.getChild("name").getContent();
    this.type = trkpt.getChild("type").getContent();
  }

}


GPX.java
/*

  Copyright (c) 2006-2011 Tom Carden

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA

*/

package tomc.gpx;

// <gpx>
//   <wpt>
//   </wpt>
//   <!-- more wpts -->
//   <trk>
//     <name>whatever</name>
//     <trkseg>
//       <trkpt lat="" lon="">
//         <ele></ele>
//         <time>yyyy-MM-ddThh:mm:ssZ</time>
//       </trkpt>
//       <!-- more trkpts -->
//     </trkseg>
//     <!-- more trksegs -->
//   </trk>
//   <!-- more trks -->
// </gpx>

import processing.core.PApplet;
import java.util.Vector;
import processing.data.XML;

/** contains a collection of GPXTrack objects and GPXWayPoint objects */
public class GPX {

  public static boolean debug = false;

  PApplet parent;

  private Vector wayPoints = new Vector();
  private Vector tracks = new Vector();

  public GPX(PApplet parent) {
    this.parent = parent;
  }

  public void parse(String url) {
    try {
      XML xmldata = parent.loadXML(url);
      for (XML xmlthing : xmldata.getChildren()) {
        if (xmlthing.getName().equals("trk")) {
          addTrack(new GPXTrack(xmlthing));
        }
        else if (xmlthing.getName().equals("wpt")) {
          addWayPoint(new GPXWayPoint(xmlthing));
        }
      }
    }
    catch(Exception e) {
      e.printStackTrace();
    }
  }

  /* Tracks */

  public void addTrack(GPXTrack trk) {
    tracks.addElement(trk);
  }

  public void removeTrack(GPXTrack trk) {
    tracks.removeElement(trk);
  }

  public GPXTrack getTrack(int i) {
    return (GPXTrack)tracks.elementAt(i);
  }

  public int getTrackCount() {
    return tracks.size();
  }

  /* WayPoints */

  public void addWayPoint(GPXWayPoint wpt) {
    wayPoints.addElement(wpt);
  }

  public void removeWayPoint(GPXWayPoint wpt) {
    wayPoints.removeElement(wpt);
  }

  public GPXWayPoint getWayPoint(int i) {
    return (GPXWayPoint)wayPoints.elementAt(i);
  }

  public int getWayPointCount() {
    return wayPoints.size();
  }

  /* Testing */

  public static void main(String[] args) {
    if (args.length >= 1) {
      GPX gpx = new GPX(new PApplet());
      GPX.debug = true;
      System.out.print("parsing " + args[0] + "... ");
      gpx.parse(args[0]);
      System.out.println("done.");
      System.out.println(gpx.getTrackCount() + " tracks parsed:");
      for (int i = 0; i < gpx.getTrackCount(); i++) {
        GPXTrack trk = gpx.getTrack(i);
        System.out.println("Track " + i + " has " + trk.size() + " track segments");
        for (int j = 0; j < trk.size(); j++) {
          GPXTrackSeg trkseg = trk.getTrackSeg(j);
          System.out.println("Segment (" + i + "," + j + ") has " + trkseg.size() + " track points");
          for (int k = 0; k < trkseg.size(); k++) {
            GPXPoint pt = trkseg.getPoint(k);
            System.out.println("Point (" + pt.lat + "," + pt.lon + ")");
          }
        }
      }
    }
    else {
      System.err.println("no filename supplied");
    }
  }

}

